iceoryx_posh/source/runtime/shared_memory_user.cpp

```diff
@@ -22,121 +22,128 @@
-        .openMode(OpenMode::OPEN_EXISTING)
-        .permissions(SHM_SEGMENT_PERMISSIONS)
-        .create()
-        .and_then([this, segmentId, segmentManagerAddressOffset](auto& sharedMemoryObject) {
-            auto registeredSuccessfully = UntypedRelativePointer::registerPtrWithId(
-                segment_id_t{segmentId},
-                sharedMemoryObject.getBaseAddress(),
-                sharedMemoryObject.get_size().expect("Failed to acquire SHM size."));
-
-            if (!registeredSuccessfully)
-            {
-                IOX_REPORT_FATAL(PoshError::POSH__SHM_APP_COULD_NOT_REGISTER_PTR_WITH_GIVEN_SEGMENT_ID);
-            }
-
-            IOX_LOG(DEBUG,
-                    "Application registered management segment "
-                        << iox::log::hex(sharedMemoryObject.getBaseAddress()) << " with size "
-                        << sharedMemoryObject.get_size().expect("Failed to acquire SHM size.") << " to id "
-                        << segmentId);
-
-            this->openDataSegments(segmentId, segmentManagerAddressOffset);
-
-            m_shmObject.emplace(std::move(sharedMemoryObject));
-        })
-        .or_else([](auto&) { IOX_REPORT_FATAL(PoshError::POSH__SHM_APP_MAPP_ERR); });
 }
 
 SharedMemoryUser::~SharedMemoryUser() noexcept
 {
-    for (auto& shm : m_dataShmObjects)
+    SharedMemoryUser::destroy(m_shmSegments);
+}
+
+void SharedMemoryUser::destroy(ShmVector_t& shmSegments) noexcept
+{
+    while (!shmSegments.empty())
     {
+        auto& shm = shmSegments.back();
         auto segmentId = segment_id_t{UntypedRelativePointer::searchId(shm.getBaseAddress())};
         UntypedRelativePointer::unregisterPtr(segmentId);
+        shmSegments.pop_back();
     }
+}
+
+expected<void, SharedMemoryUserError> SharedMemoryUser::openShmSegment(ShmVector_t& shmSegments,
+                                                                       const DomainId domainId,
+                                                                       const uint64_t segmentId,
+                                                                       const ResourceType resourceType,
+                                                                       const ShmName_t& shmName,
+                                                                       const uint64_t shmSize,
+                                                                       const AccessMode accessMode) noexcept
+{
+    constexpr access_rights SHM_SEGMENT_PERMISSIONS =
+        perms::owner_read | perms::owner_write | perms::group_read | perms::group_write;
+
+    auto shmResult = PosixSharedMemoryObjectBuilder()
+                         .name(concatenate(iceoryxResourcePrefix(domainId, resourceType), shmName))
+                         .memorySizeInBytes(shmSize)
+                         .accessMode(accessMode)
+                         .openMode(OpenMode::OPEN_EXISTING)
+                         .permissions(SHM_SEGMENT_PERMISSIONS)
```

Comment:
Can you elaborate on this? I did not change the parameters, they are the same like before.

Reply:
The permission relate to the `mode` parameter in `shm_open(const char *name, int oflag, mode_t mode);` that sets the permissions on creation when `O_CREAT` is set and the shared memory is created, otherwise it is ignored.

Reply:
Excuse my ignorance but do you suggest to just remove `.permissions(...)`?